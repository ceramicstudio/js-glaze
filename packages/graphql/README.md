# IDX GraphQL

## Installation

```sh
npm install @ceramicstudio/idx-graphql
```

## License

Apache-2.0 OR MIT