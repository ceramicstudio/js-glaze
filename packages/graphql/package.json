{
  "name": "@glazed/graphql",
  "version": "0.1.0",
  "author": "3Box Labs",
  "license": "(Apache-2.0 OR MIT)",
  "homepage": "https://github.com/ceramicstudio/js-glaze#readme",
  "keywords": [
    "ceramic",
    "glaze"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/ceramicstudio/js-glaze",
    "directory": "packages/graphql"
  },
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": "./dist/index.js"
  },
  "files": [
    "src"
  ],
  "engines": {
    "node": ">=16"
  },
  "sideEffects": false,
  "scripts": {
    "build:clean": "del dist",
    "build:js": "swc src -d ./dist --config-file ../../.swcrc",
    "build:types": "tsc --emitDeclarationOnly --skipLibCheck",
    "build": "yarn build:clean && yarn build:types && yarn build:js",
    "lint": "eslint src test --fix",
    "test": "yarn node --experimental-vm-modules $(yarn bin jest)",
    "prepare": "yarn build",
    "prepublishOnly": "package-check"
  },
  "dependencies": {
    "@ceramicnetwork/stream-tile": "^2.0.0-alpha.0",
    "@glazed/append-collection": "^0.1.0",
    "@glazed/did-datastore": "^0.2.0",
    "@glazed/graphql-types": "^0.1.0",
    "change-case": "^4.1.2",
    "graphql": "^15.7.1",
    "graphql-relay": "^0.9.0"
  },
  "devDependencies": {
    "@ceramicnetwork/common": "^2.0.0-alpha.0",
    "@ceramicnetwork/streamid": "^2.0.0-alpha.0",
    "@glazed/devtools": "^0.1.0",
    "jest-environment-glaze": "^0.1.1"
  },
  "jest": {
    "extensionsToTreatAsEsm": [
      ".ts"
    ],
    "moduleNameMapper": {
      "^(\\.{1,2}/.*)\\.js$": "$1"
    },
    "transform": {
      "^.+\\.(t|j)s$": [
        "@swc/jest",
        {
          "root": "../.."
        }
      ]
    }
  }
}
