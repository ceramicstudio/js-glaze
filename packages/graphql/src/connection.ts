import { AppendCollection } from '@glazed/append-collection'
import type { ItemResult, LoadResult } from '@glazed/append-collection'
import type { TileLoader } from '@glazed/tile-loader'
import type { Connection, ConnectionArguments, Edge } from 'graphql-relay'

import type { Doc } from './types'
import { toDoc } from './utils'

export function toConnection<T>(args: ConnectionArguments, result: LoadResult<T>): Connection<T> {
  const edges = result.items.map((item) => ({
    cursor: item.cursor.toString(),
    node: item.data,
  }))

  const firstCursor = edges[0]?.cursor
  const lastCursor = edges[edges.length - 1]?.cursor
  const pageInfo = args.first
    ? {
        startCursor: firstCursor,
        endCursor: lastCursor,
        hasNextPage: result.hasMore,
        hasPreviousPage: args.after != null,
      }
    : {
        startCursor: lastCursor,
        endCursor: firstCursor,
        hasNextPage: args.before != null,
        hasPreviousPage: result.hasMore,
      }

  return { edges, pageInfo }
}

export class ItemConnectionHandler<Node> {
  static async create<Node>(
    loader: TileLoader,
    schemaURL: string
  ): Promise<ItemConnectionHandler<Node>> {
    const collection = await AppendCollection.create<Node>(loader, schemaURL)
    return new ItemConnectionHandler(collection)
  }

  static async load<Node>(loader: TileLoader, id: string): Promise<ItemConnectionHandler<Node>> {
    const collection = await AppendCollection.load<Node>(loader, id)
    return new ItemConnectionHandler(collection)
  }

  #collection: AppendCollection<Node>

  constructor(collection: AppendCollection<Node>) {
    this.#collection = collection
  }

  get id(): string {
    return this.#collection.id.toString()
  }

  async add(node: Node): Promise<Edge<Node>> {
    const cursor = await this.#collection.add(node)
    return { cursor: cursor.toString(), node }
  }

  async load(args: ConnectionArguments): Promise<Connection<Node>> {
    let result
    if (args.first) {
      result = await this.#collection.first(args.first, args.after)
    } else if (args.last) {
      result = await this.#collection.last(args.last, args.before)
    } else {
      throw new Error('Invalid connection arguments')
    }
    return toConnection<Node>(args, result)
  }
}

export class ReferenceConnectionHandler<Node> {
  static async create<Node>(
    loader: TileLoader,
    schemaURL: string,
    nodeSchemaURL: string
  ): Promise<ReferenceConnectionHandler<Node>> {
    const collection = await AppendCollection.create<string>(loader, schemaURL)
    return new ReferenceConnectionHandler(loader, collection, nodeSchemaURL)
  }

  static async load<Node>(
    loader: TileLoader,
    id: string,
    nodeSchemaURL: string
  ): Promise<ReferenceConnectionHandler<Node>> {
    const collection = await AppendCollection.load<string>(loader, id)
    return new ReferenceConnectionHandler(loader, collection, nodeSchemaURL)
  }

  #loader: TileLoader
  #collection: AppendCollection<string>
  #nodeSchemaURL: string

  constructor(loader: TileLoader, collection: AppendCollection<string>, nodeSchemaURL: string) {
    this.#loader = loader
    this.#collection = collection
    this.#nodeSchemaURL = nodeSchemaURL
  }

  get id(): string {
    return this.#collection.id.toString()
  }

  async add(content: Node): Promise<Edge<Doc<Node>>> {
    const tile = await this.#loader.create<Node>(content, { schema: this.#nodeSchemaURL })
    const id = tile.id.toString()
    const cursor = await this.#collection.add(id)
    return { cursor: cursor.toString(), node: { id, content } }
  }

  async load(args: ConnectionArguments): Promise<Connection<Doc<Node>>> {
    let result
    if (args.first) {
      result = await this.#collection.first(args.first, args.after)
    } else if (args.last) {
      result = await this.#collection.last(args.last, args.before)
    } else {
      throw new Error('Invalid connection arguments')
    }

    const items: Array<ItemResult<Doc<Node>>> = await Promise.all(
      result.items.map(async ({ cursor, data }: ItemResult<string>) => {
        const tile = await this.#loader.load<Node>(data)
        return { cursor, data: toDoc(tile) }
      })
    )

    return toConnection(args, { items, hasMore: result.hasMore })
  }
}
