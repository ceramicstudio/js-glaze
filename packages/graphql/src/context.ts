import type { CeramicApi } from '@ceramicnetwork/common'
import { TileDocument } from '@ceramicnetwork/stream-tile'
import type { CommitID, StreamID } from '@ceramicnetwork/streamid'
import { DataModel } from '@glazed/datamodel'
import { DIDDataStore } from '@glazed/did-datastore'
import { TileLoader } from '@glazed/tile-loader'
import type { TileCache } from '@glazed/tile-loader'
import type { ModelTypeAliases, ModelTypesToAliases } from '@glazed/types'
import {
  type Connection,
  type ConnectionArguments,
  type Edge,
  cursorToOffset,
  offsetToCursor,
} from 'graphql-relay'

// import { ItemConnectionHandler, ReferenceConnectionHandler } from './connection'

const SLICE_DEFAULT_SIZE = 20

export type ContextConfig<ModelTypes extends ModelTypeAliases = ModelTypeAliases> = {
  cache?: TileCache | boolean
  ceramic: CeramicApi
  loader?: TileLoader
  model: DataModel<ModelTypes> | ModelTypesToAliases<ModelTypes>
}

export class Context<ModelTypes extends ModelTypeAliases = ModelTypeAliases> {
  #ceramic: CeramicApi
  #loader: TileLoader
  #dataStore: DIDDataStore<ModelTypes>
  // #itemConnections: Record<string, Promise<ItemConnectionHandler<unknown>>> = {}
  // #referenceConnections: Record<string, Promise<ReferenceConnectionHandler<unknown>>> = {}

  constructor(config: ContextConfig<ModelTypes>) {
    const { cache, ceramic, loader, model } = config
    this.#ceramic = ceramic
    this.#loader = loader ?? new TileLoader({ ceramic, cache })
    this.#dataStore = new DIDDataStore<ModelTypes>({
      ceramic,
      loader: this.#loader,
      model:
        model instanceof DataModel
          ? model
          : new DataModel<ModelTypes>({ loader: this.#loader, aliases: model }),
    })
  }

  get authenticated(): boolean {
    return this.#ceramic.did?.authenticated ?? false
  }

  get ceramic(): CeramicApi {
    return this.#ceramic
  }

  get dataStore(): DIDDataStore<ModelTypes> {
    return this.#dataStore
  }

  get loader(): TileLoader {
    return this.#loader
  }

  get viewerID(): string | null {
    return this.#ceramic.did?.id ?? null
  }

  // async getItemConnection<Node = unknown>(id: string): Promise<ItemConnectionHandler<Node>> {
  //   if (this.#itemConnections[id] == null) {
  //     this.#itemConnections[id] = ItemConnectionHandler.load<Node>(this.#loader, id)
  //   }
  //   return (await this.#itemConnections[id]) as ItemConnectionHandler<Node>
  // }

  // async createItemConnection<Node = unknown>(
  //   schemaURL: string
  // ): Promise<ItemConnectionHandler<Node>> {
  //   const handler = await ItemConnectionHandler.create<Node>(this.#loader, schemaURL)
  //   this.#itemConnections[handler.id] = Promise.resolve(handler)
  //   return handler
  // }

  // async getReferenceConnection<Node = unknown>(
  //   id: string,
  //   nodeSchemaURL: string
  // ): Promise<ReferenceConnectionHandler<Node>> {
  //   if (this.#referenceConnections[id] == null) {
  //     this.#referenceConnections[id] = ReferenceConnectionHandler.load<Node>(
  //       this.#loader,
  //       id,
  //       nodeSchemaURL
  //     )
  //   }
  //   return (await this.#referenceConnections[id]) as ReferenceConnectionHandler<Node>
  // }

  // async createReferenceConnection<Node = unknown>(
  //   schemaURL: string,
  //   nodeSchemaURL: string
  // ): Promise<ReferenceConnectionHandler<Node>> {
  //   const handler = await ReferenceConnectionHandler.create<Node>(
  //     this.#loader,
  //     schemaURL,
  //     nodeSchemaURL
  //   )
  //   this.#referenceConnections[handler.id] = Promise.resolve(handler)
  //   return handler
  // }

  async getRecordID(aliasOrID: string): Promise<string> {
    const definitionID = this.dataStore.getDefinitionID(aliasOrID)
    const recordID = await this.dataStore.getRecordID(definitionID)
    // Create an empty record if not already existing
    return recordID ?? (await this.dataStore.setRecord(definitionID, {})).toString()
  }

  async loadDoc<Content = Record<string, any>>(
    id: string | CommitID | StreamID,
    fresh = false
  ): Promise<TileDocument<Content>> {
    if (fresh) {
      this.#loader.clear(id)
    }
    return await this.#loader.load<Content>(id)
  }

  async createDoc<Content = Record<string, any>>(
    schema: string,
    content: Content
  ): Promise<TileDocument<Content>> {
    return await this.#loader.create<Content>(content, { schema })
  }

  async updateDoc<Content = Record<string, any>>(
    id: string | StreamID,
    content: Content
  ): Promise<TileDocument<Content | null | undefined>> {
    return await this.#loader.update(id, content)
  }

  async loadListConnection(
    list: Array<string>,
    args: ConnectionArguments = {}
  ): Promise<Connection<TileDocument | null>> {
    let sliceStart: number
    let sliceEnd: number
    if (args.last != null) {
      sliceEnd = Math.min(args.before ? cursorToOffset(args.before) : list.length, list.length)
      sliceStart = Math.max(0, sliceEnd - args.last)
    } else {
      sliceStart = Math.min(args.after ? cursorToOffset(args.after) + 1 : 0, list.length)
      sliceEnd = Math.min(sliceStart + (args.first ?? SLICE_DEFAULT_SIZE), list.length)
    }

    const sliceIDs = list.slice(sliceStart, sliceEnd)
    const results = await this.#loader.loadMany(sliceIDs)
    const edges = results.map((doc, i) => ({
      cursor: offsetToCursor(sliceStart + i),
      node: doc instanceof Error ? null : doc,
    }))

    return {
      pageInfo: {
        startCursor: edges[0]?.cursor,
        endCursor: edges[edges.length - 1]?.cursor,
        hasNextPage: sliceEnd !== list.length,
        hasPreviousPage: sliceStart !== 0,
      },
      edges: args.last ? edges.reverse() : edges,
    }
  }

  async addListConnectionEdges<T = Record<string, any>>(
    ownerID: string,
    listKey: string,
    schema: string,
    contents: Array<T>
  ): Promise<{ node: TileDocument; edges: Array<Edge<TileDocument<T>>> }> {
    const createEdgeDocs = contents.map(async (content) => await this.createDoc(schema, content))
    const [ownerDoc, ...edgeDocs] = await Promise.all([
      this.loadDoc(ownerID, true), // Force fresh load to ensure the list is up-to-date
      ...createEdgeDocs,
    ])
    const content = ownerDoc.content ?? {}
    const existing = (content[listKey] ?? []) as Array<string>

    const edgeIDs: Array<string> = []
    const edges: Array<Edge<TileDocument<T>>> = []
    edgeDocs.forEach((doc, i) => {
      edgeIDs.push(doc.id.toString())
      edges.push({ cursor: offsetToCursor(existing.length + i), node: doc })
    })
    await ownerDoc.update({ ...content, [listKey]: [...existing, ...edgeIDs] })

    return { edges, node: ownerDoc }
  }
}
