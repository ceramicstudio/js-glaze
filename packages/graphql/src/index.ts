export { GraphQLClient, type GraphQLClientParams } from './client.js'
export { Context, type ContextParams } from './context.js'
export { createGraphQLSchema } from './schema.js'
export { printGraphQLSchema } from './utils.js'
