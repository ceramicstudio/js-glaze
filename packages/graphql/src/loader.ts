import type { CeramicApi } from '@ceramicnetwork/common'
import { type CommitID, type StreamID, StreamRef } from '@ceramicnetwork/streamid'
import type { ModelInstanceDocument } from '@glazed/types'
import DataLoader from 'dataloader'
import type { BatchLoadFn } from 'dataloader'

export type DocID = CommitID | StreamID | string

// Implements CacheMap from dataloader, copied here to generate docs
export type DocumentCache = {
  /**
   * Get a Promise of a ModelInstanceDocument by its stream ID
   */
  get(id: string): Promise<ModelInstanceDocument> | void
  /**
   * Set a Promise of a ModelInstanceDocument by its stream ID
   */
  set(id: string, value: Promise<ModelInstanceDocument>): any
  /**
   * Remove a specific entry from the cache
   */
  delete(id: string): any
  /**
   * Remove all entries from the cache
   */
  clear(): any
}

export type DocumentLoaderParams = {
  /**
   * A Ceramic client instance
   */
  ceramic: CeramicApi
  /**
   * A supported cache implementation, `true` to use the default implementation or `false` to
   * disable the cache (default)
   */
  cache?: DocumentCache | boolean
  /**
   * MultiQuery request timeout in milliseconds
   */
  multiqueryTimeout?: number
}

/** @internal */
export function idToString(id: DocID): string {
  return typeof id === 'string' ? StreamRef.from(id).toString() : id.toString()
}

const tempBatchLoadFn: BatchLoadFn<DocID, ModelInstanceDocument> = () => Promise.resolve([])

export class DocumentLoader extends DataLoader<DocID, ModelInstanceDocument> {
  // #ceramic: CeramicApi
  // #useCache: boolean

  constructor(params: DocumentLoaderParams) {
    super(tempBatchLoadFn, {
      cache: true, // Cache needs to be enabled for batching
      cacheKeyFn: idToString,
      cacheMap:
        params.cache != null && typeof params.cache !== 'boolean' ? params.cache : undefined,
    })

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore internal method
    this._batchLoadFn = async (keys: ReadonlyArray<TileKey>) => {
      if (!params.cache) {
        // Disable cache but keep batching behavior - from https://github.com/graphql/dataloader#disabling-cache
        this.clearAll()
      }
      const results = await params.ceramic.multiQuery(
        keys.map(idToString).map((streamId) => ({ streamId })),
        params.multiqueryTimeout
      )
      return keys.map((key) => {
        const id = idToString(key)
        const doc = results[id]
        return doc
          ? (doc as unknown as ModelInstanceDocument)
          : new Error(`Failed to load document: ${id}`)
      })
    }

    // this.#ceramic = params.ceramic
    // this.#useCache = !!params.cache
  }

  /**
   * Add a TileDocument to the local cache, if enabled.
   */
  // cache(stream: TileDocument): boolean {
  //   if (!this.#useCache) {
  //     return false
  //   }

  //   const id = stream.id.toString()
  //   this.clear(id).prime(id, stream)
  //   return true
  // }

  /**
   * Create a new TileDocument and add it to the cache, if enabled.
   */
  // async create<T extends Record<string, any> = Record<string, any>>(
  //   content: T,
  //   metadata?: TileMetadataArgs,
  //   options?: CreateOpts
  // ): Promise<TileDocument<T>> {
  //   const stream = await TileDocument.create<T>(this.#ceramic, content, metadata, options)
  //   this.cache(stream)
  //   return stream
  // }

  /**
   * Load a ModelInstanceDocument from the cache (if enabled) or remotely.
   */
  async load<T extends Record<string, any> = Record<string, any>>(
    id: DocID
  ): Promise<ModelInstanceDocument<T>> {
    return (await super.load(id)) as ModelInstanceDocument<T>
  }

  /**
   * Update a TileDocument after loading the stream remotely, bypassing the cache.
   */
  // async update<T extends Record<string, any> = Record<string, any>>(
  //   streamID: string | StreamID,
  //   content?: T,
  //   metadata?: TileMetadataArgs,
  //   options?: UpdateOpts
  // ): Promise<TileDocument<T | null | undefined>> {
  //   const id = keyToString(streamID)
  //   this.clear(id)
  //   const stream = await this.load<T>({ streamId: id })
  //   await stream.update(content, metadata, options)
  //   return stream
  // }
}
