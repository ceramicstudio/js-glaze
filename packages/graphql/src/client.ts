import type { CeramicApi } from '@ceramicnetwork/common'
import type { DataModel } from '@glazed/datamodel'
import type { GraphQLModel } from '@glazed/graphql-types'
import type { TileCache, TileLoader } from '@glazed/tile-loader'
import type { ModelTypeAliases, ModelTypesToAliases } from '@glazed/types'
import {
  type DocumentNode,
  type ExecutionResult,
  type GraphQLError,
  type GraphQLSchema,
  type Source,
  execute,
  parse,
  validate,
} from 'graphql'

import { Context } from './context.js'
import { createGraphQLSchema } from './schema.js'

export type GraphQLClientConfig<ModelTypes extends ModelTypeAliases = ModelTypeAliases> = {
  cache?: TileCache | boolean
  ceramic: CeramicApi
  dataModel: DataModel<ModelTypes> | ModelTypesToAliases<ModelTypes>
  graphqlModel: GraphQLModel
  loader?: TileLoader
}

export class GraphQLClient<ModelTypes extends ModelTypeAliases = ModelTypeAliases> {
  #context: Context<ModelTypes>
  #schema: GraphQLSchema

  constructor(config: GraphQLClientConfig<ModelTypes>) {
    const { dataModel, graphqlModel, ...contextConfig } = config
    this.#context = new Context({ ...contextConfig, model: dataModel })
    this.#schema = createGraphQLSchema({ model: graphqlModel })
  }

  get context(): Context<ModelTypes> {
    return this.#context
  }

  async execute(
    source: string | Source,
    variableValues?: Record<string, unknown>
  ): Promise<ExecutionResult> {
    let document: DocumentNode
    try {
      document = parse(source)
    } catch (syntaxError) {
      return {
        errors: [syntaxError as GraphQLError],
      }
    }

    const validationErrors = validate(this.#schema, document)
    if (validationErrors.length > 0) {
      return {
        errors: validationErrors,
      }
    }

    return await execute({
      document,
      variableValues,
      contextValue: this.#context,
      schema: this.#schema,
    })
  }
}
