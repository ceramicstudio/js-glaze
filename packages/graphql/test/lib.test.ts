/**
 * @jest-environment glaze
 */
/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */

import type { CeramicApi } from '@ceramicnetwork/common'
import { publishCollectionSchemas } from '@glazed/append-collection'
import { ModelManager, createGraphQLModel } from '@glazed/devtools'
import { TileLoader } from '@glazed/tile-loader'
import { execute, parse, printSchema } from 'graphql'
import type { GraphQLSchema } from 'graphql'

import { Context, createGraphQLSchema } from '../src'

declare global {
  const ceramic: CeramicApi
}

describe('lib', () => {
  jest.setTimeout(20000)

  let context: Context
  let schema: GraphQLSchema

  beforeAll(async () => {
    const loader = new TileLoader({ ceramic, cache: true })

    const NoteSchema = {
      $schema: 'http://json-schema.org/draft-07/schema#',
      title: 'Note',
      type: 'object',
      properties: {
        date: {
          type: 'string',
          format: 'date-time',
          maxLength: 30,
        },
        text: {
          type: 'string',
          maxLength: 4000,
        },
        title: {
          type: 'string',
          maxLength: 100,
        },
      },
      required: ['date', 'text', 'title'],
    }
    const noteSchema = await loader.create(NoteSchema)
    const noteSchemaURL = noteSchema.commitId.toUrl()
    const noteRef = {
      type: 'string',
      title: 'reference',
      $comment: `cip88:ref:${noteSchemaURL}`,
      maxLength: 100,
    }

    const notesCollectionSchemaCommitID = await publishCollectionSchemas(loader, 'Notes', [noteRef])

    const NotesSchema = {
      $schema: 'http://json-schema.org/draft-07/schema#',
      title: 'Notes',
      type: 'object',
      properties: {
        all: {
          type: 'string',
          $comment: `cip88:ref:${notesCollectionSchemaCommitID.toUrl()}`,
          maxLength: 100,
        },
        favorites: {
          type: 'array',
          title: 'list',
          items: noteRef,
        },
      },
    }
    const notesSchema = await loader.create(NotesSchema)
    const notesSchemaURL = notesSchema.commitId.toUrl()

    const manager = new ModelManager(ceramic)
    await Promise.all([
      manager.createDefinition('myNotes', {
        name: 'notes',
        description: 'My notes',
        schema: notesSchemaURL,
      }),
      manager.createTile(
        'exampleNote',
        { date: '2020-12-10T11:12:34.567Z', text: 'An example note', title: 'Example' },
        { schema: noteSchemaURL }
      ),
    ])
    const [graphModel, model] = await Promise.all([
      createGraphQLModel(manager),
      manager.toPublished(),
    ])
    context = new Context({ ceramic, loader, model })
    schema = createGraphQLSchema(graphModel)
  })

  test('schema creation', () => {
    expect(printSchema(schema)).toMatchSnapshot()
  })

  test('read existing note', async () => {
    const readRoot = parse(`
       query TestReadNote {
         exampleNote {
           id
         }
       }
     `)
    const readNode = parse(`
       query TestReadNote($id: ID!) {
         node(id: $id) {
           ...on Note {
             date
             text
             title
           }
         }
       }
     `)

    const resRoot = await execute(schema, readRoot, {}, context)
    const resNode = await execute(schema, readNode, {}, context, {
      id: resRoot.data!.exampleNote.id,
    })
    expect(resNode).toMatchSnapshot()
  })

  test('create a note', async () => {
    const mutation = parse(`
       mutation TestCreateNote($input: CreateNoteInput!) {
         createNote(input: $input) {
           node {
             date
             text
             title
           }
         }
       }
     `)
    const res = await execute(schema, mutation, {}, context, {
      input: { content: { date: '2021-01-06T14:28:00.000Z', text: 'hello test!', title: 'test' } },
    })
    expect(res).toMatchSnapshot()
  })

  test('create, update and read a note', async () => {
    jest.setTimeout(30000)

    const create = parse(`
       mutation TestCreateNote($input: CreateNoteInput!) {
         createNote(input: $input) {
           node {
             id
           }
         }
       }
     `)
    const created = await execute(schema, create, {}, context, {
      input: { content: { date: '2021-01-06T14:28:00.000Z', text: 'hello first', title: 'first' } },
    })
    const { id } = created.data!.createNote.node

    const update = parse(`
       mutation TestUpdateNote($input: UpdateNoteInput!) {
         updateNote(input: $input) {
           node {
             id
             date
             text
             title
           }
         }
       }
     `)
    await execute(schema, update, {}, context, {
      input: {
        id,
        content: { date: '2021-01-06T14:32:00.000Z', text: 'hello second', title: 'second' },
      },
    })

    const read = parse(`
       query TestReadNote($id: ID!) {
         node(id: $id) {
           ...on Note {
             date
             text
             title
           }
         }
       }
     `)
    const res = await execute(schema, read, {}, context, { id })
    expect(res).toEqual({
      data: {
        node: { date: '2021-01-06T14:32:00.000Z', text: 'hello second', title: 'second' },
      },
    })
  })

  test('add and read notes from a connection', async () => {
    jest.setTimeout(30000)

    const create = parse(`
       mutation TestCreateNotes($input: CreateNotesInput!) {
         createNotes(input: $input) {
           node {
             id
           }
         }
       }
     `)
    const created = await execute(schema, create, {}, context, {
      input: { content: {} },
    })
    const { id } = created.data!.createNotes.node

    const add = parse(`
      mutation TestAddNoteEdge($input: AddNotesAllEdgeInput!) {
        addNotesAllEdge(input: $input) {
          edge {
            cursor
          }
        }
      }`)

    const toAdd = [
      { date: '2021-01-06T14:32:00.000Z', text: 'hello first', title: 'first' },
      { date: '2021-01-06T14:33:00.000Z', text: 'hello second', title: 'second' },
      { date: '2021-01-06T14:34:00.000Z', text: 'hello third', title: 'third' },
    ]
    for (const content of toAdd) {
      await execute(schema, add, {}, context, { input: { id, content } })
    }

    const read = parse(`
       query TestReadNotes($id: ID!) {
         node(id: $id) {
           ...on Notes {
             all(first: 3) {
               edges {
                node {
                  date
                  text
                  title
                }
               }
             }
           }
         }
       }
     `)
    await expect(execute(schema, read, {}, context, { id })).resolves.toMatchSnapshot()
  })
})
