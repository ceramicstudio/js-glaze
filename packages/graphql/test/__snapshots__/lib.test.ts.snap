// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib add and read notes from a connection 1`] = `
Array [
  Object {
    "node": Object {
      "date": "2021-01-06T14:32:00.000Z",
      "text": "hello first",
      "title": "first",
    },
  },
  Object {
    "node": Object {
      "date": "2021-01-06T14:33:00.000Z",
      "text": "hello second",
      "title": "second",
    },
  },
]
`;

exports[`lib add and read notes from a connection 2`] = `
Object {
  "data": Object {
    "node": Object {
      "allConnection": Object {
        "edges": Array [
          Object {
            "node": Object {
              "date": "2021-01-06T14:34:00.000Z",
              "text": "hello third",
              "title": "third",
            },
          },
        ],
        "pageInfo": Object {
          "hasNextPage": false,
        },
      },
    },
  },
}
`;

exports[`lib add and read notes from a connection 3`] = `
Array [
  Object {
    "node": Object {
      "date": "2021-01-06T14:34:00.000Z",
      "text": "hello third",
      "title": "third",
    },
  },
  Object {
    "node": Object {
      "date": "2021-01-06T14:33:00.000Z",
      "text": "hello second",
      "title": "second",
    },
  },
]
`;

exports[`lib add and read notes from a connection 4`] = `
Object {
  "data": Object {
    "node": Object {
      "allConnection": Object {
        "edges": Array [
          Object {
            "node": Object {
              "date": "2021-01-06T14:32:00.000Z",
              "text": "hello first",
              "title": "first",
            },
          },
        ],
        "pageInfo": Object {
          "hasPreviousPage": false,
        },
      },
    },
  },
}
`;

exports[`lib add and read notes from a connection in the store 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "store": Object {
        "notePad": Object {
          "allConnection": Object {
            "edges": Array [
              Object {
                "node": Object {
                  "date": "2021-01-06T14:32:00.000Z",
                  "text": "hello first",
                  "title": "first",
                },
              },
              Object {
                "node": Object {
                  "date": "2021-01-06T14:33:00.000Z",
                  "text": "hello second",
                  "title": "second",
                },
              },
              Object {
                "node": Object {
                  "date": "2021-01-06T14:34:00.000Z",
                  "text": "hello third",
                  "title": "third",
                },
              },
            ],
          },
        },
      },
    },
  },
}
`;

exports[`lib create a note 1`] = `
Object {
  "data": Object {
    "createNote": Object {
      "node": Object {
        "date": "2021-01-06T14:28:00.000Z",
        "text": "hello test!",
        "title": "test",
      },
    },
  },
}
`;

exports[`lib read existing note 1`] = `
Object {
  "data": Object {
    "node": Object {
      "date": "2020-12-10T11:12:34.567Z",
      "text": "An example note",
      "title": "Example",
    },
  },
}
`;

exports[`lib schema creation 1`] = `
"type Query {
  \\"\\"\\"Fetches an object given its ID\\"\\"\\"
  node(
    \\"\\"\\"The ID of an object\\"\\"\\"
    id: ID!
  ): Node
  account(id: ID!): DID!
  viewer: DID
  exampleNote: Note!
}

\\"\\"\\"An object with an ID\\"\\"\\"
interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}

type DID {
  id: ID!
  isViewer: Boolean!
  store: DataStore
}

type DataStore {
  notePad: Notes
}

type Notes implements CeramicStream & Node {
  _ceramic: CeramicMetadata!
  id: ID!
  allConnection(
    \\"\\"\\"Returns the items in the list that come after the specified cursor.\\"\\"\\"
    after: String

    \\"\\"\\"Returns the first n items from the list.\\"\\"\\"
    first: Int

    \\"\\"\\"Returns the items in the list that come before the specified cursor.\\"\\"\\"
    before: String

    \\"\\"\\"Returns the last n items from the list.\\"\\"\\"
    last: Int
  ): NoteConnection
  allNodeIDs: [ID]
  all: [String]
  favorites: [String]
}

interface CeramicStream {
  _ceramic: CeramicMetadata!
}

type CeramicMetadata {
  streamID: ID!
  commitID: ID!
  streamMetadata: StreamMetadata!
  controllers: [DID!]!
  viewerIsController: Boolean!
  schema: CeramicMetadata
}

type StreamMetadata {
  controllers: [ID!]!
  schema: String
  family: String
  tags: [String]
}

\\"\\"\\"A connection to a list of items.\\"\\"\\"
type NoteConnection {
  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"A list of edges.\\"\\"\\"
  edges: [NoteEdge]
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: String

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: String
}

\\"\\"\\"An edge in a connection.\\"\\"\\"
type NoteEdge {
  \\"\\"\\"The item at the end of the edge\\"\\"\\"
  node: Note

  \\"\\"\\"A cursor for use in pagination\\"\\"\\"
  cursor: String!
}

type Note implements CeramicStream & Node {
  _ceramic: CeramicMetadata!
  id: ID!
  date: String!
  text: String!
  title: String!
}

type Mutation {
  createNote(input: CreateNoteInput!): CreateNotePayload
  updateNote(input: UpdateNoteInput!): UpdateNotePayload
  createNotes(input: CreateNotesInput!): CreateNotesPayload
  updateNotes(input: UpdateNotesInput!): UpdateNotesPayload
  addNotesAllEdge(input: AddNotesAllEdgeInput!): AddNotesAllEdgePayload
  addNotePadNotesAllEdge(input: AddNotePadNotesAllEdgeInput!): AddNotePadNotesAllEdgePayload
  setNotePad(input: SetNotePadInput!): SetNotePadPayload
}

type CreateNotePayload {
  node: Note!
  clientMutationId: String
}

input CreateNoteInput {
  content: NoteInput!
  clientMutationId: String
}

input NoteInput {
  date: String!
  text: String!
  title: String!
}

type UpdateNotePayload {
  node: Note!
  clientMutationId: String
}

input UpdateNoteInput {
  id: ID!
  content: NoteInput!
  clientMutationId: String
}

type CreateNotesPayload {
  node: Notes!
  clientMutationId: String
}

input CreateNotesInput {
  content: NotesInput!
  clientMutationId: String
}

input NotesInput {
  all: [ID]
  favorites: [String]
}

type UpdateNotesPayload {
  node: Notes!
  clientMutationId: String
}

input UpdateNotesInput {
  id: ID!
  content: NotesInput!
  clientMutationId: String
}

type AddNotesAllEdgePayload {
  node: Notes!
  edge: NoteEdge!
  clientMutationId: String
}

input AddNotesAllEdgeInput {
  id: ID!
  content: NoteInput!
  clientMutationId: String
}

type AddNotePadNotesAllEdgePayload {
  viewer: DID!
  edge: NoteEdge!
  clientMutationId: String
}

input AddNotePadNotesAllEdgeInput {
  content: NoteInput!
  clientMutationId: String
}

type SetNotePadPayload {
  viewer: DID!
  clientMutationId: String
}

input SetNotePadInput {
  content: NotesInput!
  options: DataStoreMutationOptions
  clientMutationId: String
}

input DataStoreMutationOptions {
  merge: Boolean
}"
`;
